pkg/spire/spire_mock.go

```diff
@@ -0,0 +1,209 @@
+/*
+Copyright 2022 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spire
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// MockClient is a client used for mocking the this package for unit testing
+// other tekton components that use the spire entrypointer or controller client.
+//
```

Comment:
Hi! Sorry about being late to reply. COVID happened :|

Wanted to share a bit more about the thought process around the mocking,

---
### For unit testing

I think the client mocking will work perfectly with unit tests.

----

### For integration

We explored several different approaches when trying to mock SPIRE, a lot of the difficulty in doing so was that test that use mocked SPIRE resulted in two separate processes interacting with the shared state for integration tests. Mocking the workload API for a single memory process is ok, but when two processes (in this case the entrypointer and the controller) need to be able to retrieve compatible crypto material, is where things get a bit messy since we need to maintain state across processes and isolated containers. We would have to mock a daemon server, which pretty much re-implements a bundled version of SPIRE.

The workaround was performing mocking at a slightly higher layer, which takes away the need to have shared state between 2 processes with a fixed seed which would result in the need of an external daemon.

I think there's a desire to have the mocked SPIRE daemon with compatible fake workload client, but it is something that will be at least a couple months work of a dedicated developer.

Another possibility here could also be to create determinism within the ecdsa/rsa crypto library, it's not something I've explored too much.

---

What are some thoughts here? Maybe client mocking for unit tests but keep the higher level mocking for now for the integration/e2e tests - until the SPIRE community makes more progress on a mock server?

Reply:
No worries! Hope you're feeling better! 🙏 

That sounds good to me! My comment was more about which clients we are mocking for unit tests so that we cover more codepaths for the real implementation (in particular, I'm worried we might be missing stuff in sign.go / controller.go because spire_mock.go is using its own impl). I'd be okay with an additional higher level mock for e2e/integration tests if needed.

We may also want e2e tests that set up real spire, but I'm fine with making that out of scope for this PR.

Reply:
The e2e/integration tests already do set up the real spire. We have all this completed in the following two PRs (Phase 1 and Phase 2): https://github.com/tektoncd/pipeline/pull/4759 and https://github.com/tektoncd/pipeline/pull/4828

We are only using the mocked spire for unit testing with the various other packages (for example the pod and status).

Reply:
Oh sorry on that last "e2e" left that in there, meant to be just integration*. 

Ok so @wlynch so sounds like the plan is that we will add in the fakeworkload client for unit testing, for additional coverage and leave the alternative mocking (and use the injection mechanism instead) for integration tests? Does that sound about right?

Reply:
Sounds great to me - thanks!

Reply:
I definitely would be happy to have mocks on the workload API for unit test.
For E2E tests I would tend to say we should use the full server instead of a mock.